src/mgos_imu_icm20948.c

```diff
@@ -0,0 +1,482 @@
+    return false;
+  }
+
+  device_id = mgos_i2c_read_reg_b(i2c, i2caddr, MGOS_ICM20948_REG0_WHO_AM_I);
+  if (device_id == MGOS_ICM20948_DEVID) {
+    return true;
+  }
+  return false;
+}
+
+static bool mgos_imu_icm20948_accgyro_create(struct mgos_i2c *i2c, uint8_t i2caddr, void *imu_user_data) {
+  if (!i2c) {
+    return false;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(i2c, i2caddr, imu_user_data, 0)) {
+    return false;
+  }
+
+  // PWR_MGMNT_1: DEVICE_RESET=1; SLEEP=0; LP_EN=0; TEMP_DIS=0; CLKSEL=000;
+  mgos_i2c_write_reg_b(i2c, i2caddr, MGOS_ICM20948_REG0_PWR_MGMT_1, 0x80);
+  mgos_usleep(11000);
+
+  // PWR_MGMNT_1: DEVICE_RESET=0; SLEEP=0; LP_EN=0; TEMP_DIS=0; CLKSEL=001(auto clock source);
+  mgos_i2c_write_reg_b(i2c, i2caddr, MGOS_ICM20948_REG0_PWR_MGMT_1, 0x01);
+
+  // INT_PIN_CFG: BYPASS_EN=1(enable slave bypass mode);
+  mgos_i2c_write_reg_b(i2c, i2caddr, MGOS_ICM20948_REG0_INT_PIN_CFG, 0x02);
+
+  return true;
+}
+
+bool mgos_imu_icm20948_acc_detect(struct mgos_imu_acc *dev, void *imu_user_data) {
+  return mgos_imu_icm20948_detect(dev->i2c, dev->i2caddr, imu_user_data);
+}
+
+bool mgos_imu_icm20948_acc_create(struct mgos_imu_acc *dev, void *imu_user_data) {
+  struct mgos_imu_icm20948_userdata *iud = (struct mgos_imu_icm20948_userdata *)imu_user_data;
+
+  if (!dev) {
+    return false;
+  }
+
+  // Only initialize the ICM20948 if gyro hasn't done so yet
+  if (!iud->accgyro_initialized) {
+    if (!mgos_imu_icm20948_accgyro_create(dev->i2c, dev->i2caddr, imu_user_data)) {
+      return false;
+    }
+    iud->accgyro_initialized = true;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+
+  // REG2_ACCEL_CONFIG: ACCEL_DLPFCFG=101(12Hz); ACCEL_FS_SEL=10(8g); ACCEL_FCHOICE=1(Enable accel DLPF);
+  // ACCEL_SMPLRT_DIV_1: ACCEL_SMPLRT_DIV=0000(MSB);
+  // ACCEL_SMPLRT_DIV_2: ACCEL_SMPLRT_DIV=00001010(LSB);
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_CONFIG, 0x15);
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_SMPLRT_DIV_1, 0x00);
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_SMPLRT_DIV_2, 0x0a);
+
+  dev->scale = 8.f / 32767.0f;
+  return true;
+}
+
+bool mgos_imu_icm20948_acc_read(struct mgos_imu_acc *dev, void *imu_user_data) {
+  uint8_t data[6];
+
+  if (!dev) {
+    return false;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 0)) {
+    return false;
+  }
+  if (!mgos_i2c_read_reg_n(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG0_ACCEL_XOUT_H, 6, data)) {
+    return false;
+  }
+
+  dev->ax = (data[0] << 8) | (data[1]);
+  dev->ay = (data[2] << 8) | (data[3]);
+  dev->az = (data[4] << 8) | (data[5]);
+
+  return true;
+}
+
+bool mgos_imu_icm20948_acc_get_scale(struct mgos_imu_acc *dev, void *imu_user_data, float *scale) {
+  uint8_t fs = 0;
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
```

Comment:
Please ensure that *scale is a valid pointer before trying to write to it later (causing a crash).
if (!scale) return false;